src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs

```diff
@@ -207,51 +207,108 @@ private void SignalCompletion()
+                    ExceptionDispatchInfo? edi = InvokeInlineContinuation();
+                    // Restore the current ExecutionContext.
+                    ExecutionContext.Restore(currentContext);
+                    // Now rethrow the exception; if there is one.
+                    edi?.Throw();
                 }
+
+                return;
             }
+
+            InvokeSchedulerContinuation();
+            // Restore the current ExecutionContext.
+            ExecutionContext.Restore(currentContext);
         }
 
         /// <summary>
-        /// Invokes the continuation with the appropriate captured context / scheduler.
-        /// This assumes that if <see cref="_executionContext"/> is not null we're already
+        /// Invokes the continuation inline and captures any exception thrown.
+        /// This assumes that if <see cref="_continuation"/> is not null we're already
         /// running within that <see cref="ExecutionContext"/>.
         /// </summary>
-        private void InvokeContinuation()
+        private ExceptionDispatchInfo? InvokeInlineContinuation()
         {
+            // This is in a helper as the error handling causes the generated asm
+            // for the surrounding code to become less efficent (stack spills etc)
+            // and it is an uncommon path.
+
             Debug.Assert(_continuation != null);
+            Debug.Assert(_capturedContext == null);
+            Debug.Assert(!RunContinuationsAsynchronously);
 
-            switch (_capturedContext)
+            ExceptionDispatchInfo? edi = null;
+            SynchronizationContext? syncContext = SynchronizationContext.Current;
+            try
             {
-                case null:
-                    if (RunContinuationsAsynchronously)
-                    {
-                        if (_executionContext != null)
-                        {
-                            ThreadPool.QueueUserWorkItem(_continuation, _continuationState, preferLocal: true);
-                        }
-                        else
-                        {
-                            ThreadPool.UnsafeQueueUserWorkItem(_continuation, _continuationState, preferLocal: true);
-                        }
-                    }
-                    else
-                    {
-                        _continuation(_continuationState);
-                    }
-                    break;
+                _continuation(_continuationState);
+            }
+            catch (Exception ex)
+            {
+                // Note: we have a "catch" rather than a "finally" because we want
+                // to stop the first pass of EH here.  That way we can restore the previous
+                // context before any of our callers' EH filters run.
+                edi = ExceptionDispatchInfo.Capture(ex);
+            }
+
+            // Set sync context back to what it was prior to coming in
+            SynchronizationContext.SetSynchronizationContext(syncContext);
```

Comment:
This should probably be in a finally block attached to the above try/catch (if necessary at all based on my other comment)

Reply:
Either the exception is stopped by the catch all or its an exception that can't be stopped when the SynchronizationContext doesn't really matter?

Reply:
`ExceptionDispatchInfo.Capture()` may also throw OOM since it allocates, restoring in finally should cover that

Reply:
Probably in catastrophic failure state at that point; but changed to be in finally blocks